Implement block header

rtsisyk opened the pull request:
- Add stubs for PublicKey, Signature, Payload and other
- Define Input, Output, BlockHeader and Block structures
- Implement block signature
- Implement genesis block in simplified way

Closes #32

dbmcclain commented:
I implemented needed items in `crypto::pbc`
1. New dummy constructor for `crypto::pbc::secure::RVal` used in encrypted payloads, printable, hash-able
2. Implement `crypto::pbc::secure::Signature` type - printable, hash-able
3. make internal routines Hash::from_vector() and Hash::from_str() both utilize the new Hasher object and Hashable.
4. Make crypto::Hash items re-hash-able
5. Implement error handling where needed
6. Base our hex strings around the hex crate functions

dbmcclain commented:
Not sure what you are thinking here about EncryptedPayload. That is not a crypto function, but something specific to a particular application.

The pbc::secure module defines a complete structure called EncryptedPacket that includes a public key, an ID tag, the RVal, and the crypto-text. This is a generic SAKKI encrypted output that can be used by applications as they see fit.

Our blockchain is just one particular application for SAKKI encryption, and it only needs to show the RVal and crypto-text portions. The public key shall remain hidden, known already by the intended recipient, and also same for the ID tag.

I argue that this kind of specificity for blockchain does not belong in a general purpose crypto library, and so the definition of EncryptedPayload belongs with the client code that cherry picks what it needs from a SAKKI EncryptedPacket structure.

rtsisyk commented:
> Not sure what you are thinking here about EncryptedPayload. That is not a crypto function, but something specific to a particular application.

OK, I agree with you. I moved EncryptedPayload back to blockchain library. Definition is from your comment above.

> As for the missing libPBC in Linux... I am at a loss to help you there. Why did it work previously and not now?

This problem was caused by a bug in our CI scripts. This has been already fixed.

Rebased and force-pushed. Please review.

dbmcclain commented on blockchain/src/output.rs:
```diff
@@ -0,0 +1,82 @@
+
+impl Hashable for Output {
+    fn hash(&self, state: &mut Hasher) {
+        // Don't include self.hash because it is redundant in this case
```
... self.hash not just redundant - ID should equal the result of this call to hash. You might want to check that condition to be sure you aren't dealing with garbage data from some forgery.

rtsisyk replied:
This function is used when Output itself is hashed, i.e. for create the block signature.
There are no difference between hash(hash(a, b, c), a, b, c) and hash(a, b, c) in terms of signing.

dbmcclain replied:
yes... there really is a difference... but maybe you don't need to care?

dbmcclain commented on crypto/src/lib.rs:
```diff
@@ -25,17 +25,42 @@ extern crate hex;
+/// Stub for Bullet Proof
+// TODO: define
+#[derive(Clone)]
+pub struct BulletProof([u8; 4096]);
```
yes, as good a guess as any. Should flag this with TODO to come back and fix once we know what a Bulletproof actually looks like. (Actually, I'm the one that will likely do this...)

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -0,0 +1,226 @@
+
+    /// CoSi multisignature on HCURR
+    // TODO: which kind
+    pub sig: G1,
```
OK

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -0,0 +1,226 @@
+        leader: PublicKey,
+        adjustment: Zr,
+        witnesses: &[PublicKey],
+        sig: G1,
```
OK

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -0,0 +1,226 @@
+        for witness in witnesses.iter() {
+            witness.hash(&mut hasher);
+        }
+        sig.hash(&mut hasher);
```
OK, I understood the idea. I'll rework this part.

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -0,0 +1,226 @@
+        let previous = Hash::from_str(GENESIS_BLOCK_HASH_STRING);
+        let adjustment: Zr = Zr::new();
+        let witnesses = [leader.clone()];
+        let sig: G1 = G1::new();
```
I'll rework it today.

rtsisyk commented on blockchain/src/block.rs:
```diff
@@ -0,0 +1,226 @@
+
+        Block::sign(
+            version, epoch, previous, leader, adjustment, &witnesses, sig, &inputs, &outputs,
+        )
```
OK

rtsisyk commented on blockchain/src/input.rs:
```diff
@@ -0,0 +1,51 @@
+    /// Identifier of an unspent transaction output.
+    pub source_id: Hash,
+    /// Signature used to sign transactions.
+    pub signature: Signature,
```
I guessed that Input should be signed by sender...

rtsisyk commented:
I'm reworking it according to David's comments.
